ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+      *)
+      Option.iter (fun log -> undo_changes !log) log;
+      Error e
+
+  let submode obj a b =
+    match submode_try obj a b with
+    | Ok log ->
+      Option.iter !append_changes log;
+      Ok ()
+    | Error _ as e -> e
+
+  let zap_to_ceil_morphvar obj mv =
+    assert (submode obj (Amode (mupper obj mv)) (Amodevar mv) |> Result.is_ok);
+    mupper obj mv
+
+  let zap_to_ceil : type a l. a C.obj -> (a, l * allowed) mode -> a =
+   fun obj -> function
+    | Amode m -> m
+    | Amodevar mv -> zap_to_ceil_morphvar obj mv
+    | Amodemeet (a, mvs) ->
+      List.fold_left
+        (fun acc mv -> C.meet obj acc (zap_to_ceil_morphvar obj mv))
+        a mvs
+
+  let join (type a r) obj l =
+    let rec loop :
+        a ->
+        (a, allowed * disallowed) morphvar list ->
+        (a, allowed * r) mode list ->
+        (a, allowed * disallowed) mode =
+     fun a mvs ->
+      if C.le obj (C.max obj) a
+      then fun _ -> Amode (C.max obj)
+      else
+        function
+        | [] -> Amodejoin (a, mvs)
+        | mv :: xs -> (
+          match disallow_right mv with
+          | Amode b -> loop (C.join obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
+          | Amodejoin (b, mvs') -> loop (C.join obj a b) (mvs' @ mvs) xs)
+    in
+    loop (C.min obj) [] l
+
+  let meet (type a l) obj l =
+    let rec loop :
+        a ->
+        (a, disallowed * allowed) morphvar list ->
+        (a, l * allowed) mode list ->
+        (a, disallowed * allowed) mode =
+     fun a mvs ->
+      if C.le obj a (C.min obj)
+      then fun _ -> Amode (C.min obj)
+      else
+        function
+        | [] -> Amodemeet (a, mvs)
+        | mv :: xs -> (
+          match disallow_left mv with
+          | Amode b -> loop (C.meet obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
+          | Amodemeet (b, mvs') -> loop (C.meet obj a b) (mvs' @ mvs) xs)
+    in
+    loop (C.max obj) [] l
+
+  let zap_to_floor_morphvar ~commit obj mv =
+    let log, lower = constrain_mlower_try obj mv in
+    if commit then !append_changes log else undo_changes !log;
+    lower
+
+  let zap_to_floor : type a r. a C.obj -> (a, allowed * r) mode -> a =
+   fun obj -> function
+    | Amode a -> a
+    | Amodevar mv -> zap_to_floor_morphvar ~commit:true obj mv
+    | Amodejoin (a, mvs) ->
+      List.fold_left
+        (fun acc mv ->
+          C.join obj acc (zap_to_floor_morphvar ~commit:true obj mv))
+        a mvs
+
+  (* because lower bound conservative, this check is also conservative.
+     if it returns Some, then definitely a constant.
+     if it returns None, then we don't know anything *)
+  let check_const : type a l r. a C.obj -> (a, l * r) mode -> a option =
+   fun obj -> function
+    | Amode a -> Some a
+    | Amodevar mv ->
+      let lower = zap_to_floor_morphvar ~commit:false obj mv in
+      if C.le obj (mupper obj mv) lower then Some lower else None
+    | Amodemeet (a, mvs) ->
```

Comment:
Some comments here in this next section might help.
* The upper bound of a meet mode is the meet of the upper bounds of the components. 
* The lower bound of a meet mode is the meet of the lower bounds of the components.

Those definitions are symmetric.... but the implementations are not. Why?

Reply:
I think what you wrote here is correct - the implementations are not symmetric, again because `mupper` is precise but `mlower` is not (and hence we need to call `zap_to_floor` to get the precise lower bound)